File: Data Structures & Algorithms/longest-common-subsequence/submission-5.java
class Pair{
    private int i;
    private int j;

    public Pair(int i,int j){
        this.i=i;
        this.j=j;
    }

    public int getI(){
        return this.i;
    }
    public int getJ(){
        return this.j;
    }   
    @Override 
    public boolean equals(Object p){
        if(this == p) return true;
        if (p == null || getClass() != p.getClass()) return false;
        Pair pair = (Pair) p;
        return i == pair.i && j == pair.j;
    }
    @Override
    public int hashCode() {
        return Objects.hash(i, j);
    }
}
class Solution {
    public Map<Pair,Integer> d = new HashMap<>();
    public int dfs(String text1,String text2,int i,int j , int n ,int m){
        if(d.containsKey(new Pair(i,j))){
            return d.get(new Pair(i,j));
        }
        int res=0;
        if(i==n || j==m) return 0;
        if(text1.charAt(i) == text2.charAt(j)){
           res= 1+dfs(text1,text2,i+1,j+1,n,m);
           d.put(new Pair(i,j),res);
           return res ;
        }
        res= Math.max(dfs(text1,text2,i+1,j,n,m),dfs(text1,text2,i,j+1,n,m));
        d.put(new Pair(i,j),res);
        return res ;
    }
    public int longestCommonSubsequence(String text1, String text2) {
        return dfs(text1,text2,0,0,text1.length(),text2.length());
    }
}

File: Data Structures & Algorithms/edit-distance/submission-5.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if word1 == word2 :
            return 0
        n,m=len(word1),len(word2)
        if min(n,m)==0:
            return max(n,m)
        d={}
        def dfs(i,j):
            res=0
            if (i,j) in d :
                return d[(i,j)]
            if word1[i]==word2[j] :
                if j == m-1 and i == n-1 :
                    return 0
                elif j==m-1 :
                    res += n-i-1
                elif i==n-1 :
                    res+= m-j-1
                else:
                    res+=dfs(i+1,j+1)
            else:
                res+=1
                if i==n-1 and j==m-1 :
                    return res
                elif i==n-1 :
                    res+=min(dfs(i,j+1),m-j-1)
                elif j==m-1 :
                    res+=min(dfs(i+1,j),n-i-1)
                else:
                    res += min(dfs(i,j+1) , dfs(i+1,j) ,dfs(i+1,j+1))
            d[(i,j)]=res
            return res
        return dfs(0,0)